RFC 54: Internationalisation of Wagtail content

kaedroho opened the pull request:
This RFC proposes an approach for implementing basic support for internationalised content in Wagtail.

[Rendered](https://github.com/wagtail/rfcs/blob/internationalisation/text/054-internationalisation.md)

This is sponsored by the Mozilla Foundation.

allcaps commented:
Ping @Henk-JanVanHasselaar and @mikedingjan. Since you are the authors of [wagtailtrans](https://github.com/wagtail/wagtailtrans). Can you please have a look at this RFC? Do you have feedback for us?

kaedroho commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+If `WAGTAIL_I18N_ENABLED` is `False`. All new pages at the root will use the locale that is set in the `LANGUAGE_CODE` setting.
+
+#### Editing pages
```
Need to mention that the page editor may be affected by the language of the page being edited. Eg, an Arabic page will trigger all fields to right-align.

zerolab replied:
Does that mean this overrides the user preferences? (i.e. `admin/account/language_preferences/`) Say my preferred language is English, but I can do Arabic translation. Should the interface change? What do other systems do?

kaedroho replied:
Your interface will still be English, but the fields on the Arabic pages would be right-aligned in order to display the text correctly. If your interface language is set to Arabic and you edit an English page, the text will be left-aligned.

I'm pretty sure this is the only way we can make this work, I'm not sure if Arabic is even readable if it's left-aligned.

kaedroho replied:
I've added a paragraph abou this into the RFC now

ababic commented:
@kaedroho  **Question**: I'm a little unclear on the need for the `unique_together` changes on `Page`. Whatever code created translations would have to catch/raise integrity errors anyhow, so might it good to leave things open and let the workflow implementation take care of checking/enforcing?

kaedroho commented:
> Whatever code created translations would have to catch/raise integrity errors anyhow, so might it good to leave things open and let the workflow implementation take care of checking/enforcing?

I think it's very unlikely that you would ever get an integrity error here, and we already do the checks to avoid them (like we won't create a translation if there already is a page there).

The constraints are there in case there's ever a race condition, which are almost impossible to avoid without database constraints. One past example is we didn't used to have a `unique_together` constraint on `wagtailimages.Rendition` (on the image/filter-spec fields) and this led to some people getting duplicate renditions in their database even though Wagtail uses `get_or_create`!

ababic commented:
@kaedroho **Suggestion**: It could be handy to have `translated()` and `not_translated()` methods on `PageQuerySet` - no reason these cannot be added later though.

allcaps commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+### Locales and Languages
+
+This RFC proposes to add a `Locale` model with a `language_code` field. The locales are defined by a setting called `WAGTAIL_LANGUAGES`. This looks inconsistent, I know. But locales have a different meaning to languages:
```
We have `LANGUAGES`, `WAGTAIL_LANGUAGES` doesn't a locale model duplicate the same data? Why can't a Page store the language code as locale identifier (char field)? What is the advantage of a model?

kaedroho replied:
I went with a model mainly because it's simpler to change your language codes globally. For example, you might have an `en` locale and you later want to split that to `en-GB` and `en-US` as your website grows.

There's a couple of other minor reasons:

 - It's nice to have a class that you can add methods on. For example `page.locale.get_display_name()`
 - We might want to add new fields to this later

But both of these have good alternatives that don't require this `Locale` model.

It's definitely a problem in Wagtail Localize at the moment that you have to manually run `sync_languages` whenever you change the `WAGTAIL_LANGUAGES` setting. I haven't tried the idea for syncing this automatically on startup (as mentioned in the RFC) yet.

kaedroho replied:
Django CMS uses a `CharField` for this: https://github.com/divio/django-cms/blob/develop/cms/models/pagemodel.py#L223

RFC 9 and `wagtailtrans` both have a language model. But it looks like they're intended to be managed by an admin which I'm not going for here (I'd like developers to decide which languages can be used on a site).

kaedroho replied:
I wonder if we change this slightly: Rather than Wagtail localize trying to stay in perfect sync with this setting, we allow administrators to define locales themselves, but the locales they set up must be in this list.

If a locale is removed from this list but is still being used, we display a warning in the admin and let them either assign a new language or disable it.

allcaps replied:
Good arguments.

Keeping code and database in sync isn't very nice, but most of the time better than alternatives. I do like if these locale model instances are defined in code (the setting), I don't like the threw the web interface for manual sync. As you said, this is for the developer to handle.

Creating missing locale entries on startup and leave orphan locales for the developer to clean up seems sensible.

jjanssen replied:
One of the main reasons for wagtailtrans to choose for a Language model is that we didn't want be dependent on code deployments or database migrations whenever one of our clients wants to add a new language.

To put this a bit in perspective: one of our clients has a multisite environment containing 32 sites and each site has a range of languages with a minimum of at least 2 and I have seen a max of 8. Per site there can be overlapping languages but it can be completely different languages as well. Wagtailtrans has the ability to configure languages per site as well.

As this client is experienced and their application manager is able to handle this aspect for their companies worldwide, than it wouldn't make sense for us to have to this managed by a developer or let them have to call us every time they needed a new language for one of their sites.

allcaps commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+When `WAGTAIL_I18N_ENABLED` is `True`, a new “Translations” menu item will appear at the top of the page editor:
+
+![](https://paper-attachments.dropbox.com/s_447CE1D894BBD41C78540EBB08A4706330CCCE70BE0BB00C4C6B8C7EFCD9AF76_1594118046048_image.png)
```
The changes in Page layout where also shown during the Torchbox webinar. Some page elements need to be repositioned to free screen estate for translation ui elements. This is all fine. But moving elements to a new position when you toggle WAGTAIL_I18N_ENABLED isn't smart.

For example the Page type label moves from prepending the page title (WAGTAIL_I18N_ENABLED=False) to below the page title (WAGTAIL_I18N_ENABLED=True). This means adding Page type label two times to the template and keeping track of `WAGTAIL_I18N_ENABLED` setting.

I think it is wise to keep elements the same place. That way they don't need to be wrapped in conditions.
That way development, maintenance, browser testing but also user adaptation is easier.

zerolab replied:
@allcaps the Page type label will move under the title as part of the workflow work. With `WAGTAIL_I18N_ENABLED=True`, the "Translations" item will get appended to that row, so the should be no positional changes

kaedroho replied:
The position of the page type/status will be moved underneath the title in Wagtail 2.10.

I'm thinking we could just add the translations link here as it's also where workflow status and page history is listed and I think it's similar.

allcaps commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+We will prevent editors from moving pages to a parent that has a different locale to itself.
+
+Open question: Should we just change the locale of the page instead?
```
Prevent moving is okay. Just notify the user.

gasman replied:
I'd slightly lean towards yes (unless there's an existing translation in that locale, as that would break the unique constraint). If a user tries to do that we should try to make a best guess at what they intended to do, and I think that's the only reasonable interpretation.

kaedroho replied:
If there are any internal links to that page, these wouldn't be updated if we allowed a move so this could potentially lead to accidental cross-linking between sites. But I guess we can always warn the user about that.

kaedroho commented:
> Do we need to add something about how Wagtail localisation works on regular Django views?

Since we're using `i18n_patterns`, you can just add your Django views into that as you would on any Django app. This was done for the search view on the demo here: https://localize-workflow.herokuapp.com/en/search/

The only thing I had to do differently was add a check for the current active language code and filter the pages.

KalobTaulien commented:
Nice and thorough RFC, I’ve read through it and like the approach you’re taking. 

I think this would drastically simplify i18n with wagtail sites, and that’s a huge win for Wagtail in general since i18n can be a difficult feature for people to implement in general.

kaedroho commented:
I think we should change the logic of `max_count` to allow you to create that number of a page type in each locale. https://docs.wagtail.io/en/stable/reference/pages/model_reference.html#wagtail.core.models.Page.max_count

For example, this might be set to `1` for a homepage, but if we didn't take locales into account, we wouldn't be able to translate anything because we need to create a new site tree. It wouldn't be possible to move pages between the trees so this wouldn't create any loopholes.

Any thoughts?

kaedroho commented:
I've been looking into how we could adjust the routing logic to account for locales.

I think there's two main options:

 1) *one site one locale* - Add a "locale" field to site and set up different site records for each locale. These site records would share a domain name and port but would have a different locale. We then add some logic into `Site.find_for_request` function to check the requests current language and fetch the correct site.
 2) *one site many locales* - Create one site and link it to the English version. Then add logic into the serve view to look for a translation of the root page if the request locale is not English. We will also need to override the URL generation logic to account for the root page being in a different tree.

Wagtail localize used option 2, but I'm starting to like option 1 more now, it requires a lot less logic to be overridden (sites can still use `site.root_page` too).

kaedroho commented:
I'm thinking "one site one locale" isn't feasible because of redirects and settings.

Redirects shouldn't be required to be created in a single locale, since they are often used to reserve URLS from a migrated system that might've not had a locale prefix. If you want a locale prefix, you can add this into the redirect manually.

Settings might not be locale-specific (eg, a google analytics key). But there should be the option to make locale-specific settings by adding a `TranslatableMixin` to the model.

So I think there's two options:
1) **Magic approach** - Link your `site_root` to the English version of the site (or whatever language is specified in `LANGUAGE_CODE`). We will work out the other site roots by looking for translations
2) **Explicit approach** - Make `site_root` repeatable and require a user to manually set up the pages (but they really should be translations of each other in order for some logic to work as expected

lb- commented:
Hey. This is incredibly thorough and well thought out, I have minimal experience with j18n systems so cannot really give much input. I may have missed it in the detail, but how does the model know that Page X is the same as Page Y (but just a different language). If each page has its own UUID, how do they get 'linked' to say they are the same content. 
... Or is this what it's up to other libraries to implement?

kaedroho commented:
Hey @lb- thanks for looking at it!

Pages that are the same but in different languages have the same `translation_key` UUID. So to find translations of a page, you have to filter pages that have the same UUID.

lb- commented:
Ok. And will Wagtail have the UI added to make a Page share the same translation key?

Also, does this mean, in theory, you could have a Page share a translation key with a snippet? (E.g. a custom setup might have snippets serve as their non-canonical translations)

kaedroho commented:
> Ok. And will Wagtail have the UI added to make a Page share the same translation key?

The `copy_for_translation` method would handle setting the UUID correctly. There will be no UI exposed for this method though, the UI for making and managing translations would remain in third party modules for now. This is an area we haven't explored much yet and I think it's best to try out a few ways in third party modules before settling on one.

> Also, does this mean, in theory, you could have a Page share a translation key with a snippet? (E.g. a custom setup might have snippets serve as their non-canonical translations)

Two completely different models could share a translation key, but this wouldn't mean anything. At the database level, the snippets would have a completely different `translation_key`/`locale` fields to `Page`.

It is possible to have two different page types be translations of each other though, which shouldn't be an issue for the changes proposed here but might cause problems in a third-party module that needs them to be the same. But as the third-party module would be responsible for creating the translated pages I think this is fine.

lb- commented:
Thanks @kaedroho - appreciate the explanations, sounds good.

gasman commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+* **Page queries are simplified** - you don’t need to filter out other languages from `get_children` or `get_siblings`. Also, you can perform any query from any page and get correct results (running `get_children` from a translation will never return results)
+* **Permissions and Workflows can be configured per-language** - Wagtail permissions and workflows are both hooked into the tree and inherited. Using a separated tree approach allows translators to be grouped by their locale and only given permission to pages in that locale
+* **Different locales can be structured differently** - I’m not sure why people would want to purposefully structure different language sections in different ways. But the main advantage to this is it allows page structure to be updated asynchronously.
```
I can think of a couple of possibilities (although they don't really change the arguments here):

* Different languages have vastly different volumes of content. If your recipe site has 1000 recipes in English organised by category, but only 10 have been translated into Swahili, the English version will probably have a deeper hierarchy.
* The content in different languages is adapted for cultural or geographical considerations.

gasman commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+    ('en-US', "English (United States)"),
+    ('es-ES', "Spanish (Spain)"),
+    ('es-MX', "Spanish (Mexico)"),
+]
```
Presumably we're OK with (and would encourage) people using translated strings here? `_("English (United Kingdom)")`

kaedroho replied:
Yep, I'll update the example (sorry, was lazy)

gasman commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+* `get_translation(locale)` - Returns a translated object if one exists in the specified locale, otherwise raises `model.DoesNotExist`
+* `get_translation_or_none(locale)` - Returns a translated object if one exists in the specified locale, otherwise returns `None`
+* `has_translation(locale)` - Returns a boolean if a translation for this object exists in the specified locale.
+* `copy_for_translation(locale)` - Generates a copy of the object with the same content and `translation_key` but a different `locale`
```
I see you've accounted for this being a special case for Page, which is good... hopefully anyone building translatable non-page models that need extra logic on copy (e.g. updating a slug with a unique constraint on it) will also be aware of the need to override copy_for_translation. However, one common case where this might not be obvious is snippet / modeladmin models using ClusterableModel / InlinePanel. Perhaps TranslatableMixin.copy_for_translation needs to recognise when it's being called on a ClusterableModel and deal with copying the child objects?

kaedroho replied:
Oh yes, that's a good point!

kaedroho replied:
Mentioned this in RFC now so we shouldn't forget to do this

gasman commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+#### Overridden methods
+
+* `copy_for_translation(locale, copy_parents=False)`  - Overridden to be based on `Page.copy()` but also implement a `copy_parents` flag. When set to `True` this will copy any parents that are not translated yet in order for the structure to match the source tree. If the parent doesn’t exist and `copy_parents` is `False`, this method would raise a `ParentNotTranslated` exception.
```
Does it publish the copied page?

kaedroho replied:
No, it doesn't

gasman commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+#### Page explorer
+
+We will add a language switcher into the React page explorer. At the root, this filters the pages by language. When navigating into a sub page, this could be used to switch to the page in other language trees (so the possible values would be limited by what languages the current page is translated into).
```
Need to figure out how this interacts with permissions - presumably the language list should only include languages for which there is at least one existing page they have edit permission on. Limiting the language list according to the current page seems a bit constraining - wouldn't it be more intuitive to keep the language list fixed, and on switching language, go to the closest ancestor for which a translation is available?

gasman commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+* Display a text value with a link in the “translation status” column (if this hook is not used, that column will be hidden)
+* Add extra action buttons next to each translation
+* Add global actions at the bottom of the modal (For example: “Translate this page into another language”)
```
Given that there's no way to actually create translations without installing a third-party package, would it not be cleaner to make the whole modal the responsibility of the third-party package? That way, Wagtail core only has to provide a hook for inserting the "n translations" link on the edit view.

kaedroho replied:
Yep, I prefer this, thanks!

kaedroho commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+
+Advantages of **separated tree** are:
+
+* **Page queries are simplified** - you don’t need to filter out other languages from `get_children` or `get_siblings`. Also, you can perform any query from any page and get correct results (running `get_children` from a translation will never return results)
```
Sorry, this is poorly worded. That sentence is describing combined trees where the translations are made as siblings of the English version. This means that all children of a translated page would actually be made as children in the English tree. This could really do with an image to help visualise it!

kaedroho commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+* **Permissions and Workflows can be configured per-language** - Wagtail permissions and workflows are both hooked into the tree and inherited. Using a separated tree approach allows translators to be grouped by their locale and only given permission to pages in that locale
+* **Different locales can be structured differently** - I’m not sure why people would want to purposefully structure different language sections in different ways. But the main advantage to this is it allows page structure to be updated asynchronously.
+
+The models in this RFC could, in theory, work with both of these approaches. But allowing both approaches will complicate some of the logic so we will only support the **separated tree** approach for now.
```
I think there are a few places where we would need to update the logic to account for this:

 - `copy_for_translation` would need to be able to copy pages into either structure
 - Routing would need to be implemented in different ways.
  - for separated trees, the routing would be done at the site root page (so we check if the site root page has a translation in the current language and use that translation as the site root instead), this allows slugs to be different for different languages
  - for combined trees, the routing would all be done in English. Once it's found the page in English, it'll look for a translation
 - Because routing would work differently, generating URLs to pages would need to work differently too
 - Moving pages - separated trees would have logic to prevent pages from being moved from one language to another. Not sure how this would work in a combined trees though

gasman commented:
> I think we should change the logic of `max_count` to allow you to create that number of a page type in each locale. https://docs.wagtail.io/en/stable/reference/pages/model_reference.html#wagtail.core.models.Page.max_count
> 
> For example, this might be set to `1` for a homepage, but if we didn't take locales into account, we wouldn't be able to translate anything because we need to create a new site tree. It wouldn't be possible to move pages between the trees so this wouldn't create any loopholes.
> 
> Any thoughts?

This makes sense, yes. I think there may be some edge cases that come up around page trees that 'accidentally' have a mixture of locales due to a developer switching LANGUAGE_CODE partway through development (will expand on this in a separate comment shortly...) but I think these probably only come up on multi-site installations, and I suspect that `Page.max_count` doesn't serve any sensible logical purpose on those anyhow. :-)

gasman commented:
Thinking through the implications of a developer casually changing LANGUAGE_CODE partway through development while WAGTAIL_I18N_ENABLED is unset - which is something that a Django developer would reasonably expect to be able to do without major consequences...

Any project created with the Wagtail project template will initially have LANGUAGE_CODE set to en-us. If they _don't_ immediately change this prior to running `./manage.py migrate` - and maybe even if they do, depending on how clever the `home/0002_create_homepage` migration is going to be - then the homepage will be created with locale=en-us.

Supposing the developer then changes LANGUAGE_CODE to nl, and proceeds to build their (Dutch-language) website underneath that homepage - all subpages will inherit the locale=en-us of the homepage. This is fine, because the locale field isn't exposed to the user anywhere while WAGTAIL_I18N_ENABLED is unset.

They then create a new homepage at the root level - this will be created with locale=nl. Again, this is fine, as long as we're careful to ensure that any locale-aware functionality in Wagtail (such as filtering the explorer menu to only show pages in the active locale, or disallowing users from moving pages to a destination with a different locale) is only applied when WAGTAIL_I18N_ENABLED is true.

Eventually, they decide to enable multi-language support, and find that it's reporting the sites under two different locales (and thus need to explicitly switch between them when navigating in the explorer, for example).

So, as far as I can see, the practical implications are:

* Wagtail needs to explicitly check WAGTAIL_I18N_ENABLED on all locale-aware code, and not rely on non-i18n-enabled sites having the same locale for all pages
* wagtailtrans and wagtail-localize should probably provide the ability to bulk-update the locale field on a subtree, to fix that issue where it comes up.

kaedroho commented:
@gasman That's an interesting point, thanks!

Following the discussion in https://github.com/wagtail/rfcs/pull/54#discussion_r452421766, I've been thinking we should add a UI to let admins set up locales but restrict them to the options in `WAGTAIL_LANGUAGES` rather than creating them automatically.

If we did this, there would be a crash when you attempted to create that new homepage because the `Locale` wouldn't exist. Which would be good in that you'll never be in a situation where you have multiple locales being used on a non-i18n site, but bat in that i18n shouldn't be breaking things for non-i18n users.

I think we could change the default here. if `WAGTAIL_I18N_ENABLED` is `False`, it always takes the locale of the parent page, even if that parent is at the root. If they later decide to enable `I18n`, the locale's `language_code` would be wrong, but all pages would share the same `Locale` object so this could be updated on that record rather than having to change a lot of pages `locale`.

Does that sound OK?

kaedroho commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,530 @@
+* **It doesn't require any magic to insert fields into models** - duplicating the fields can get unwieldy if done manually. So usually some sort of meta-class has to be used to add them. This can make page models harder to work with.
+* **Page slugs can also be translated** - Wagtail only has one slug field and this can’t be duplicated without customising the way Wagtail’s routing works.
+
+The main disadvantage to this approach is non-translated fields that you might want to remain the same for all languages will be duplicated for each language. But this is a translation management concern so this is left to the third-party module to solve (`wagtail-localize` has solved this).
```
Yep, they can. Wagtail localize updates non-translated fields when any translations from the same version of the page get applied. This prevents inconsistencies like an image caption and image not matching each other.

kaedroho commented:
Hey Everyone,

Thanks for all the really quick and detailed reviews!

I've just made some updates which clarifies some areas and makes the following changes:

 - Renamed `WAGTAIL_LANGUAGES` to `WAGTAIL_CONTENT_LANGUAGES` to make it clear that these languages are not for the Wagtail UI
 - Locales are now explicitly created through the admin (but options limited by `WAGTAIL_CONTENT_LANGUAGES`). This is to address issues that could occur if language codes were changed over time
 - I've added sections on routing an page.max_count (the routing is already implemented in https://github.com/wagtail/wagtail/pull/6220
 - I've made some changes to the UI and API sections as well
 - I've clarified that the UI section can change significantly after the RFC is merged, to allow us to make further design improvements

jjanssen commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,579 @@
+
+#### Editing pages
+
+When `WAGTAIL_I18N_ENABLED` is `True`, a new language switcher will appear at the top of the page editor. We will implement a hook to allow third-party apps to add their own buttons here too. A translation app could then hook in any extra UI they need, such as displaying translation progress for example.
```
Do we also want to provide this for usage in the public templates? I can imagine if this mechanism will be available in the admin there could be a general interest in a possibility to use this in public html as well.

kaedroho replied:
Yep, people would definitely want a language switcher in their frontends. But I don't think Wagtail needs to provide this as it's already quite simple to implement in the template. For example https://github.com/wagtail/bakerydemo/commit/7d84b87d83813d7c750e5c273b52396c352cee34#diff-38c4c0d8249f9721b1c114492c6ce6a2R23-R26

I think we can provide this snippet in the documentation

kaedroho commented:
One new thing I'm thinking of adding to this is a `.local` property to all translatable objects. It'll be similar to `.specific` but will look up a version of the page/snippet in the currently activated language.

The main use for this is where you copy a page but don't update all of the foreign keys to other pages (for example, a featured blog post on a homepage). But you still want to use a localised version if there is one.

In this case, you can do this in your template:
```html+Django
{% with page.featured_blog_post.local as blog_post %}
    <a href="{% pageurl blog_post %}">{{ blog_post.title }}</a>
{% endwith %}
```

If that blog post is translated into the active language, they will get a link to the local version. Otherwise, the version that's actually in the database will be returned (so they might be linked to a page in a different language in this case)

I'd also like to make this the default behaviour of internal links in rich rich text fields.

Let me know if anyone has any thoughts on that

gasman commented:
@kaedroho Sounds like a useful addition, but maybe call it `.localized`? `.local` leaves open a lot of possibilities for what it means to someone reading the code - as a developer I'd probably assume it was something to do with syncing between local and remote databases.

kaedroho commented on text/054-internationalisation.md:
```diff
@@ -0,0 +1,579 @@
+#### Fields
+
+* `language_code` (string, unique) - This is a language code taken from the list in `WAGTAIL_CONTENT_LANGUAGES`. Locales are defined by their language. The language code must be a [BCP 47 language tag](https://tools.ietf.org/html/bcp47) (the same format used in Django).
+* `is_active` (boolean) - Rather than deleting locales when they are removed form `WAGTAIL_CONTENT_LANGUAGES`, this boolean is set to `False` preventing new content from being translated into this `Locale`. It’s up to the developer to clean up their database.
```
Going to remove this field in favour of a filter in the manager. By default, `Locale.objects.all()` will only return locales that are currently in `WAGTAIL_CONTENT_LANGUAGES`. Any removed locales can be accessed with `Locale.all_objects.all()` (only the locale management UI will use this though).

kaedroho commented:
@gasman that makes sense, thanks!

kaedroho commented:
Moving to final comment period as this is mostly implemented now and the implementation closely matches what is described in the RFC.
